broker.go

```diff
@@ -165,6 +165,17 @@ func (b *Broker) Open(conf *Config) error {
 
 		if conf.Net.TLS.Enable {
 			Logger.Printf("Using tls")
+
+			// If no ServerName is set, infer the ServerName
+			// from the hostname we're connecting to.
+			if conf.Net.TLS.Config.ServerName == "" {
```

Comment:
That's going to segfault in the default case where the Sarma user has set `conf.Net.TLS.Enable = true` but hasn't change `conf.Net.TLS.Config` from the default nil pointer value

To restore the old behaviour you need to add a:
```go
cfg := conf.Net.TLS.Config
if cfg == nil {
    cfg = &tls.Config{}
    // then proceeed to set ServerName 
}

```

Reply:
See my updated comment on the PR.
Not sure if my change + your proposal (which I agree) will be enough

Reply:
currently master doesn't do a TLS handshake per `tls.Client`, this makes me nervous :/

Reply:
scratch that.
https://pkg.go.dev/crypto/tls?tab=doc#Conn.Handshake

> Handshake runs the client or server handshake protocol if it has not yet been run. Most uses of this package need not call Handshake explicitly: the first Read or Write will call it automatically.

Reply:
I just pushed the config change. Still need to figure out how to add a test for this.

The big difference now is that we aren't doing the timeouts/deadlines and handshake checks on this specific point, unsure what would be the effect of that.

Reply:
I did modify one test to remove the `ServerName`, that might be good enough for now https://github.com/Shopify/sarama/pull/1692/files#diff-e9534989c7143940469fa20938dbcb71

what do you think @dnwe ?

Reply:
👍🏻